feat(experiments): MDE Calculation with historical data.

rodrigoi opened the pull request:
> [!IMPORTANT]
> 👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Problem

Our existing running time calculator uses the `$feature_flag_called` event to suggest the experiment's size and duration, and the MDE is hard coded to 5%. This PR allows the user to use a specific event (typically a `$pageview`) as the first element in the series, thus allowing the user to use real historical data to calculate the experiment parameters. 

This PR also includes some refactors and bug fixes, but the important changes are limited to a few files.

<!-- Who are we building for, what are their needs, why is this important? -->

<!-- Does this fix an issue? Uncomment the line below with the issue ID to automatically close it when merged -->
<!-- Closes #ISSUE_ID -->

## Changes

### Tidying Up Some Code.

The largest tidying is on the `RunningTimeCalculatorModal` component, where we extracted `MeanMetricDataPanel` and `FunnelMetricDataPanel` into their own components without code changes.

We also created some new components:
- `RunningTimeCalculatorModalStep` is a new component that reuses code to create the step definition.
- `RunningTimeCalculatorModalFooter` contains the... modal footer. Close and Save functions are provided as props.
- `EventSelectorStep` and `MetricSelectorStep` encapsulate the functionality of the two first steps, which coincidentally are the ones that interact with server data.

This PR also includes two small additions to the type system: the `isExperimentMeanMetric` and `isExperimenttFunnelMetric` type guards, which help make type narrowing more declarative.

### Logic and d Calculation Changes

These are the most important changes to the code, mostly in the `runningTimeCalculatorLogic.tsx` file.

To support multivariate experiments with several funnel steps, we modified `getEventField` to return the last element in the `series`. This allows us to calculate the conversion rate correctly for the entire series, not just the first two elements.
Generally, when dealing with series, we always select the first and last element, even if the return results have only two elements. This will future-proof the code/

The baseline event configuration is stored on the global state and saved to the database on the Experiment `properties`. This required small changes in the logic types and reducers.

The suggested MDE is also stored in state and calculated when `metricResult` completes. This required changes on the reducer and loader.
We also refactored `metricResult` to make the return statement more legible. We are not abusing the spread operator that much, and there's one duplicated line of code, but the goal is to improve the legibility of the code.

### UI Updates

| Before | After |
| - | - |
| <img width="704" alt="image" src="https://github.com/user-attachments/assets/47c7a110-6276-4c12-ab3b-8daf31fd2086" /> | <img width="706" alt="image" src="https://github.com/user-attachments/assets/1ff4573a-bd28-43a2-b7cb-741850620e11" /> |

### Bug Fix?

| <img width="678" alt="image" src="https://github.com/user-attachments/assets/b98b5b21-84a3-4f0f-b0c0-67f150c1e4b0" /> |
| - |
| Backend Error when querying for Funnel Metrics |

When executing the `FunnelsQuery`, the backend complains that the `kind` of the first element in the series was wrongly set to `NodeKind. `FunnelsQuery`. We modified `getKindField` to use `EventNode` instead, which apparently fixed the problem.


<!-- If there are frontend changes, please include screenshots. -->
<!-- If a reference design was involved, include a link to the relevant Figma frame! -->

## Does this work well for both Cloud and self-hosted?

<!-- Yes / no / it doesn't have an impact. -->

## How did you test this code?

<!-- Briefly describe the steps you took. -->
<!-- Include automated tests if possible, otherwise describe the manual testing routine. -->

rodrigoi committed:
refactor: extracting event selector step.

rodrigoi committed:
refactor: extract the funnel metric data panel component.

rodrigoi committed:
refactor: extracting the mean metrics data panel component.

rodrigoi committed:
refactor: extracting the modal footer component.

rodrigoi committed:
feat: adding event configuration to running time modal logic.

rodrigoi committed:
refactor: extracting the modal step component.

rodrigoi committed:
refactor: extracting the metric selector step.

rodrigoi committed:
feat: adding event configuration to experiment properties.

rodrigoi committed:
chore: remove console log statements from metrics data panels.

rodrigoi committed:
refactor: cleaning up event selector step.

rodrigoi committed:
feat: adding event configuration to the running time calculator modal.

rodrigoi committed:
feat: adding type guards for metric types and selecting the last event in the series for funnels.

rodrigoi committed:
feat: modifiying MDE calculation.

rodrigoi committed:
feat: updating the copy of the first step.

andehen reviewed (commented):
Code looks great to me! Well structured into components and they only care about the data they have to. Overall, I think the end-user experience here is a lot better than the current version and actually yield a useful estimation on running time. Good job! 🙌 

A couple of reflections:
- we should update the title and descriptions to make it clear that we are asking for something that can estimate traffic to the experiment
- if I just want to go with the pre-selected `$pageview`, it's not clear how I can proceed to the next step as those are hidden. It just looks like this then
<img width="761" alt="Screenshot 2025-04-04 at 14 54 21" src="https://github.com/user-attachments/assets/aa08d106-ec9e-4551-a7b6-9e18d82b3811" />

So that can be a little confusing then. Maybe we should also show step two by default?

- minor, but would be nice if we populated it with the data from db when one clicks "calculate" after a page refresh. Now one starts on scratch again.
- we should default to something like 30 % MDE. That's a more reasonable starting point I think. Otherwise, users will get very long running times by default. Also, 1 % is a granular enough step size for that one. I also can't erase the last digit in the form, so it's a little annoying to check with keyboard input.

I also realize now that there is actually no reason to just use the last step of the funnel as I suggested. We can use the whole series to get the most accurate conversion rate. But we can fix that in follow up later if we want to.

rodrigoi committed:
fix: applying greptile suggestion for better checks on EventSelectorStep.

rodrigoi committed:
fix: guarding against division by zero on CR count.

rodrigoi commented on frontend/src/scenes/experiments/RunningTimeCalculator/runningTimeCalculatorLogic.tsx:
```diff
@@ -211,27 +239,42 @@ export const runningTimeCalculatorLogic = kea<runningTimeCalculatorLogicType>([
+                if (isExperimentFunnelMetric(metric)) {
+                    const automaticConversionRateDecimal =
+                        result?.results?.at(-1)?.count / result?.results?.at(0)?.count || null
+
```
Rookie mistake fixed on fc9499cf5e42fa3480f8c1544e2a8460beb71873.

rodrigoi committed:
fix: updating first step copy.

rodrigoi commented on frontend/src/scenes/experiments/RunningTimeCalculator/EventSelectorStep.tsx:
```diff
@@ -0,0 +1,56 @@
+    return (
+        <RunningTimeCalculatorModalStep
+            stepNumber={1}
+            title="Select Baseline Event"
```
Updated on 70dd1a3cee0b4e33ac1a52cac1b94716abdc5ad9.

<img width="703" alt="image" src="https://github.com/user-attachments/assets/94873572-45bc-4c06-a87d-bc80c1e6f82e" />

rodrigoi commented on frontend/src/scenes/experiments/RunningTimeCalculator/EventSelectorStep.tsx:
```diff
@@ -0,0 +1,56 @@
+        <RunningTimeCalculatorModalStep
+            stepNumber={1}
+            title="Select Baseline Event"
+            description="Choose an event to establish your baseline metrics. We'll analyze its historical data to calculate the minimum sample size and estimated duration for your experiment."
```
Updated on 70dd1a3cee0b4e33ac1a52cac1b94716abdc5ad9.
<img width="703" alt="image" src="https://github.com/user-attachments/assets/bf35c6c0-8b74-4f62-90f8-5ecffa2e7f68" />

rodrigoi committed:
fix: user can select a metric with  as default.

rodrigoi committed:
fix: changing the step on the MDE box.

rodrigoi committed:
fix: updating default MDE to 30%.

rodrigoi commented:
@andehen here is the list of fixes:

- Copy updated on 70dd1a3cee0b4e33ac1a52cac1b94716abdc5ad9.
- Default state won't block the metrics dialog on 772478519d88220c3ff7545b7f4f5fffe226c0df
- Changed the step on the MDE form field here 4ce56ffe1092e398293b0ee7cfe14ab16c838643 and changed the default to `30%` here 773ee9c7ea27e3748a1449646940f391aefe7b8a.

I'm working on recovering the state for "edit" functionality, but that's looking like a different PR.

| Updated UI |
| - |
| <img width="706" alt="image" src="https://github.com/user-attachments/assets/82246e8c-46a6-4479-80a6-3c0542409631" /> |

andehen commented on frontend/src/scenes/experiments/RunningTimeCalculator/runningTimeCalculatorLogic.tsx:
```diff
@@ -211,27 +245,45 @@ export const runningTimeCalculatorLogic = kea<runningTimeCalculatorLogicType>([
                           metric.source.math === ExperimentMetricMathType.Sum
                         ? getSumQuery(metric, values.experiment)
-                        : getFunnelQuery(metric, values.experiment)
+                        : getFunnelQuery(metric, values.eventConfig, values.experiment)
```
Typo
```suggestion
                        : getFunnelQuery(metric, values.exposureEstimateConfig, values.experiment)
```

andehen commented on frontend/src/scenes/experiments/utils.ts:
```diff
@@ -97,7 +97,7 @@ export function getMinimumDetectableEffect(
         const mde = Math.ceil(targetConversionRate - currentConversionRate)
 
-        return mde || 5
+        return mde || 30
```
Just a side note for later: I think we can get rid of this whole function actually. It think this calculation tries to estimate an absolute MDE, but we should just default to 30% relative MDE. So no need for this logic. But that's another PR.

rodrigoi replied:
added to my todo list 😉

andehen commented on frontend/src/scenes/experiments/RunningTimeCalculator/runningTimeCalculatorLogic.tsx:
```diff
@@ -124,13 +134,18 @@ const getSumQuery = (metric: ExperimentMetric, experiment: Experiment): TrendsQu
             {
                 kind: NodeKind.EventsNode,
-                event: '$feature_flag_called',
+                event: eventConfig?.event ?? '$feature_flag_called',
```
```suggestion
                event: eventConfig?.event ?? '$pageview',
```

rodrigoi replied:
This made sense to me in the case of an experiment with a reused feature flag. But I agree that `$pageview` is the right default here. Maybe we can make this smarter by checking the featue flag age.

Fixed here 8bf2ec8598d3c0f46a46be4754394e4aa269d842.

andehen commented on frontend/src/scenes/experiments/RunningTimeCalculator/RunningTimeCalculatorModal.tsx:
```diff
@@ -140,110 +30,39 @@ export function RunningTimeCalculatorModal(): JSX.Element {
+                        <RunningTimeCalculatorModalStep
+                            stepNumber={3}
+                            title="Choose minimum detectable effect"
+                            description="The minimum detectable effect (MDE) is the smallest improvement you want to be able to detect with statistical significance. A smaller MDE requires more participants but can detect subtler changes."
```
We should make it clear here that it's _relative_ MDE and not absolute (just a suggestion, feel free to reword)
```suggestion
                            description="The minimum detectable effect (MDE) is the smallest relative improvement you want to be able to detect with statistical significance. A smaller MDE requires more participants but can detect subtler changes."
```

rodrigoi replied:
I went through several iterations of this copy before setting on this one. I'm surprised you only requested one change 😛 

Updated here: 3cfa05fe3ef0477d7a1aa90529636930e616f64f

andehen reviewed (commented):
I guess it's still some WIP, but had a closer now now anyways. This is starting to look really good! 👍 Feels very good to have it separated into components like this.

Tested (with the suggestions made) and I think the UX is pretty good! The results make sense and it feels very useful.

We should also add a listener like this so the results are updated when the first step is changed
```ts
    listeners(({ actions }) => ({
        setMetricIndex: () => {
            actions.loadMetricResult()
        },
        setExposureEstimateConfig: () => { // <- This is new
            actions.loadMetricResult()
        },
    })),
```

rodrigoi committed:
fix: type errors.

rodrigoi committed:
feat: updating metric results when the exposure estimate changes.

rodrigoi commented:
@andehen ! Updated the listeners here: 6a10365ce4772162c85ce50790c979735fd26563 (thanks, I'm still fighting Kea)

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

rodrigoi committed:
fix: updating defaults on FunnelQuery to .

rodrigoi committed:
Merge branch 'experiments/running-time-calculator-update' of github.com:PostHog/posthog into experiments/running-time-calculator-update

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

rodrigoi committed:
fix: updating the copy for the MDE calculator.

rodrigoi committed:
Merge branch 'experiments/running-time-calculator-update' of github.com:PostHog/posthog into experiments/running-time-calculator-update

rodrigoi committed:
feat: updating the queries schema to support new experiment config.

rodrigoi committed:
fix: updating Python schema mappings to TS.

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

andehen commented on frontend/src/scenes/experiments/RunningTimeCalculator/RunningTimeCalculatorModal.tsx:
```diff
@@ -140,110 +30,39 @@ export function RunningTimeCalculatorModal(): JSX.Element {
+                        <RunningTimeCalculatorModalStep
+                            stepNumber={3}
+                            title="Choose minimum detectable effect"
+                            description="The minimum detectable effect (MDE) is the smallest relativeimprovement you want to be able to detect with statistical significance. A smaller MDE requires more participants but can detect subtler changes."
```
Missing space
```suggestion
                            description="The minimum detectable effect (MDE) is the smallest relative improvement you want to be able to detect with statistical significance. A smaller MDE requires more participants but can detect subtler changes."
```

andehen commented on frontend/src/scenes/experiments/RunningTimeCalculator/runningTimeCalculatorLogic.tsx:
```diff
@@ -62,8 +67,13 @@ const getEventField = (metric: ExperimentMetric): string | number | null | undef
+         * For multivariate funnels, we select the last step
+         * Although we know that the last step is always an EventsNode, TS infers that the last step might be undefined
+         * so we use the non-null assertion operator (!) to tell TS that we know the last step is always an EventsNode
+         */
```
Thanks for adding comments like this!

andehen reviewed (approved):
Looks good ! 👍  Great job on this.

I noticed one issue in that we allow metric math to be `undefined`, where we default to `TotalCount`. However, we don't handle it well here in this logic. But I think the right fix is to not allow `undefined` at all and default to `TotalCount` when saving metrics. Thus, we don't have to deal with `undefiend` values all over the place. Nothing to change here then.

![1](https://github.com/user-attachments/assets/339d3ae8-9f89-43bd-91ab-6ef837c489a3)

andehen committed:
Update frontend/src/scenes/experiments/RunningTimeCalculator/RunningTimeCalculatorModal.tsx

rodrigoi committed:
Merge branch 'master' into experiments/running-time-calculator-update
